{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c0ba0df6-85ff-4dde-b3cd-6b05e81e3e6a",
   "metadata": {},
   "source": [
    "# Import statements"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3de0596d-7a3b-4d0f-84d2-8e3c563f5556",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "from helpers import *\n",
    "\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "84d356d2-fd99-4014-975b-30a2823bcdd3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "env: PATH=/usr/bin:/home/youn/mambaforge/envs/chronostrain2/bin\n"
     ]
    }
   ],
   "source": [
    "\"\"\" ============================================ DO NOT EDIT BELOW ============================================ \"\"\"\n",
    "\"\"\" environment variable extraction \"\"\"\n",
    "try:\n",
    "    VARS_SET\n",
    "except NameError:\n",
    "    VARS_SET = True\n",
    "    _cwd = %pwd\n",
    "    _parent_cwd = Path(_cwd).parent\n",
    "    _start_path = %env PATH\n",
    "        \n",
    "%env PATH=/usr/bin:$_start_path"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56e516ff-065a-4b83-b8af-3ba652666db8",
   "metadata": {},
   "source": [
    "# Important global vars\n",
    "\n",
    "IMPORTANT to edit these!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "238c86b8-80c1-4c60-9724-9d9110f80990",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_basedir = Path().resolve().parent.parent / \"data\" / \"chronostrain_seeds\" / \"Escherichia_coli\"  # the directory to store the gene files and marker index file.\n",
    "refseq_metadata_description = \"Escherichia coli K12 MG1655\"  # This goes into the metadata of the FASTA records.\n",
    "mlst_id_prefix = \"E_Coli\"\n",
    "\n",
    "\n",
    "reference_genome_fasta = Path() / \"E_Coli_Reference.fasta\"\n",
    "assert reference_genome_fasta.exists()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72f91e8b-03e2-420f-aa2c-b99562f86cd8",
   "metadata": {},
   "source": [
    "# Extract MLST typing genes."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e180b7d7-0dc0-4a1c-a9b7-beae003ac02c",
   "metadata": {},
   "source": [
    "As of 2024 Sept. 9th, E. coli has the Pasteur+Actman scheme. https://pubmlst.org/bigsdb?db=pubmlst_mlst_seqdef"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "29032c7a-0c6e-4f34-801b-6d7339dd0e9b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>adk</td>\n",
       "      <td>Escherichia_coli_1_adk.fasta</td>\n",
       "      <td>MLST:Escherichia_coli_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>fumC</td>\n",
       "      <td>Escherichia_coli_1_fumC.fasta</td>\n",
       "      <td>MLST:Escherichia_coli_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gyrB</td>\n",
       "      <td>Escherichia_coli_1_gyrB.fasta</td>\n",
       "      <td>MLST:Escherichia_coli_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>icd</td>\n",
       "      <td>Escherichia_coli_1_icd.fasta</td>\n",
       "      <td>MLST:Escherichia_coli_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>mdh</td>\n",
       "      <td>Escherichia_coli_1_mdh.fasta</td>\n",
       "      <td>MLST:Escherichia_coli_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>purA</td>\n",
       "      <td>Escherichia_coli_1_purA.fasta</td>\n",
       "      <td>MLST:Escherichia_coli_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>recA</td>\n",
       "      <td>Escherichia_coli_1_recA.fasta</td>\n",
       "      <td>MLST:Escherichia_coli_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>dinB</td>\n",
       "      <td>Escherichia_coli_2_dinB.fasta</td>\n",
       "      <td>MLST:Escherichia_coli_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>icdA</td>\n",
       "      <td>Escherichia_coli_2_icdA.fasta</td>\n",
       "      <td>MLST:Escherichia_coli_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>pabB</td>\n",
       "      <td>Escherichia_coli_2_pabB.fasta</td>\n",
       "      <td>MLST:Escherichia_coli_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>polB</td>\n",
       "      <td>Escherichia_coli_2_polB.fasta</td>\n",
       "      <td>MLST:Escherichia_coli_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>putP</td>\n",
       "      <td>Escherichia_coli_2_putP.fasta</td>\n",
       "      <td>MLST:Escherichia_coli_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>trpA</td>\n",
       "      <td>Escherichia_coli_2_trpA.fasta</td>\n",
       "      <td>MLST:Escherichia_coli_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>trpB</td>\n",
       "      <td>Escherichia_coli_2_trpB.fasta</td>\n",
       "      <td>MLST:Escherichia_coli_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>uidA</td>\n",
       "      <td>Escherichia_coli_2_uidA.fasta</td>\n",
       "      <td>MLST:Escherichia_coli_2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       0                              1                        2\n",
       "0    adk   Escherichia_coli_1_adk.fasta  MLST:Escherichia_coli_1\n",
       "1   fumC  Escherichia_coli_1_fumC.fasta  MLST:Escherichia_coli_1\n",
       "2   gyrB  Escherichia_coli_1_gyrB.fasta  MLST:Escherichia_coli_1\n",
       "3    icd   Escherichia_coli_1_icd.fasta  MLST:Escherichia_coli_1\n",
       "4    mdh   Escherichia_coli_1_mdh.fasta  MLST:Escherichia_coli_1\n",
       "5   purA  Escherichia_coli_1_purA.fasta  MLST:Escherichia_coli_1\n",
       "6   recA  Escherichia_coli_1_recA.fasta  MLST:Escherichia_coli_1\n",
       "7   dinB  Escherichia_coli_2_dinB.fasta  MLST:Escherichia_coli_2\n",
       "8   icdA  Escherichia_coli_2_icdA.fasta  MLST:Escherichia_coli_2\n",
       "9   pabB  Escherichia_coli_2_pabB.fasta  MLST:Escherichia_coli_2\n",
       "10  polB  Escherichia_coli_2_polB.fasta  MLST:Escherichia_coli_2\n",
       "11  putP  Escherichia_coli_2_putP.fasta  MLST:Escherichia_coli_2\n",
       "12  trpA  Escherichia_coli_2_trpA.fasta  MLST:Escherichia_coli_2\n",
       "13  trpB  Escherichia_coli_2_trpB.fasta  MLST:Escherichia_coli_2\n",
       "14  uidA  Escherichia_coli_2_uidA.fasta  MLST:Escherichia_coli_2"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "paper_marker_dir = Path().resolve().parent.parent / \"data\" / \"chronostrain_seeds\" / \"Escherichia_coli_custom\"\n",
    "paper_df = pd.read_csv(paper_marker_dir / \"marker_seed_index.tsv\", sep='\\t', header=None)\n",
    "paper_mlst_slice = paper_df.loc[paper_df[2].str.startswith(\"MLST\")]\n",
    "paper_mlst_slice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "739f8682-61a3-400e-a219-5c0ca0dc87d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Fasta</th>\n",
       "      <th>Metadata</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>adk</td>\n",
       "      <td>adk.fasta</td>\n",
       "      <td>MLST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>fumC</td>\n",
       "      <td>fumC.fasta</td>\n",
       "      <td>MLST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gyrB</td>\n",
       "      <td>gyrB.fasta</td>\n",
       "      <td>MLST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>icd</td>\n",
       "      <td>icd.fasta</td>\n",
       "      <td>MLST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>mdh</td>\n",
       "      <td>mdh.fasta</td>\n",
       "      <td>MLST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>purA</td>\n",
       "      <td>purA.fasta</td>\n",
       "      <td>MLST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>recA</td>\n",
       "      <td>recA.fasta</td>\n",
       "      <td>MLST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>dinB</td>\n",
       "      <td>dinB.fasta</td>\n",
       "      <td>MLST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>icdA</td>\n",
       "      <td>icdA.fasta</td>\n",
       "      <td>MLST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>pabB</td>\n",
       "      <td>pabB.fasta</td>\n",
       "      <td>MLST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>polB</td>\n",
       "      <td>polB.fasta</td>\n",
       "      <td>MLST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>putP</td>\n",
       "      <td>putP.fasta</td>\n",
       "      <td>MLST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>trpA</td>\n",
       "      <td>trpA.fasta</td>\n",
       "      <td>MLST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>trpB</td>\n",
       "      <td>trpB.fasta</td>\n",
       "      <td>MLST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>uidA</td>\n",
       "      <td>uidA.fasta</td>\n",
       "      <td>MLST</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Name       Fasta Metadata\n",
       "0    adk   adk.fasta     MLST\n",
       "1   fumC  fumC.fasta     MLST\n",
       "2   gyrB  gyrB.fasta     MLST\n",
       "3    icd   icd.fasta     MLST\n",
       "4    mdh   mdh.fasta     MLST\n",
       "5   purA  purA.fasta     MLST\n",
       "6   recA  recA.fasta     MLST\n",
       "7   dinB  dinB.fasta     MLST\n",
       "8   icdA  icdA.fasta     MLST\n",
       "9   pabB  pabB.fasta     MLST\n",
       "10  polB  polB.fasta     MLST\n",
       "11  putP  putP.fasta     MLST\n",
       "12  trpA  trpA.fasta     MLST\n",
       "13  trpB  trpB.fasta     MLST\n",
       "14  uidA  uidA.fasta     MLST"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import shutil\n",
    "\n",
    "mlst_index = []\n",
    "for _, row in paper_mlst_slice.iterrows():\n",
    "    gene_name = row[0]\n",
    "    gene_path = paper_marker_dir / row[1]\n",
    "    filename = f'{gene_name}.fasta'\n",
    "    shutil.copy(gene_path, gene_basedir / filename)\n",
    "    mlst_index.append({'Name': f'{gene_name}', 'Fasta': filename, 'Metadata': 'MLST'})\n",
    "mlst_index = pd.DataFrame(mlst_index)\n",
    "display(mlst_index)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a703bbd-8caf-455c-ba40-69c2d3716def",
   "metadata": {},
   "source": [
    "# Extract MetaPhlAn markers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "daf4cedb-85ea-459b-89bd-5f608a44c6c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "taxon_label = 's__Escherichia_coli'\n",
    "metaphlan_version_id = \"mpa_vJun23_CHOCOPhlAnSGB_202307\"\n",
    "metaphlan_pkl_path = Path(f\"/mnt/e/metaphlan_databases/{metaphlan_version_id}/{metaphlan_version_id}.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "64c34dcd-00f8-442e-bd30-c3bffa94892f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_from_metaphlan(input_metaphlan_pkl: Path):\n",
    "    parser = MetaphlanParser(input_metaphlan_pkl)\n",
    "\n",
    "    # Extract reference seqs\n",
    "    metaphlan_gene_records = []\n",
    "    for marker_name, record in parser.retrieve_marker_seeds(taxon_label):\n",
    "        marker_len = len(record.seq)\n",
    "        print(f\"Found marker `{marker_name}` (length {marker_len})\")\n",
    "        metaphlan_gene_records.append(record)\n",
    "    return metaphlan_gene_records"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "7436597b-3c68-4607-b17a-e3aba6e98477",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Searching for marker seeds from MetaPhlAn database: mpa_vJun23_CHOCOPhlAnSGB_202307.\n",
      "Target # of markers: 136\n",
      "Found marker `UniRef90_P75933|1__4|SGB10068` (length 600)\n",
      "Found marker `UniRef90_Q0T2M6|4__9|SGB10068` (length 800)\n",
      "Found marker `UniRef90_P33916|5__8|SGB10068` (length 500)\n",
      "Found marker `UniRef90_P25718|1__4|SGB10068` (length 550)\n",
      "Found marker `UniRef90_Q8XAU9|1__3|SGB10068` (length 450)\n",
      "Found marker `UniRef90_P52614|1__4|SGB10068` (length 550)\n",
      "Found marker `UniRef90_P25798|4__8|SGB10068` (length 650)\n",
      "Found marker `UniRef90_P0AFI1|2__6|SGB10068` (length 650)\n",
      "Found marker `UniRef90_P07017|2__6|SGB10068` (length 650)\n",
      "Found marker `UniRef90_B7UQY2|3__6|SGB10068` (length 500)\n",
      "Found marker `UniRef90_P21362|1__3|SGB10068` (length 450)\n",
      "Found marker `UniRef90_Q47141|1__5|SGB10068` (length 750)\n",
      "Found marker `UniRef90_P0ABR6|1__9|SGB10068` (length 1350)\n",
      "Found marker `UniRef90_P76254|1__5|SGB10068` (length 700)\n",
      "Found marker `UniRef90_Q8XA71|5__8|SGB10068` (length 550)\n",
      "Found marker `UniRef90_P0AEW2|1__4|SGB10068` (length 600)\n",
      "Found marker `UniRef90_P77744|1__6|SGB10068` (length 900)\n",
      "Found marker `UniRef90_P0ABW6|1__3|SGB10068` (length 450)\n",
      "Found marker `UniRef90_P45759|6__9|SGB10068` (length 550)\n",
      "Found marker `UniRef90_P39338|3__7|SGB10068` (length 700)\n",
      "Found marker `UniRef90_P13656|1__7|SGB10068` (length 1000)\n",
      "Found marker `UniRef90_P10805|1__5|SGB10068` (length 750)\n",
      "Found marker `UniRef90_P31131|1__3|SGB10068` (length 450)\n",
      "Found marker `UniRef90_P50466|2__7|SGB10068` (length 800)\n",
      "Found marker `UniRef90_P76584|1__5|SGB10068` (length 750)\n",
      "Found marker `UniRef90_P77334|10__13|SGB10068` (length 550)\n",
      "Found marker `UniRef90_P77334|2__8|SGB10068` (length 950)\n",
      "Found marker `UniRef90_P69814|1__3|SGB10068` (length 450)\n",
      "Found marker `UniRef90_P77256|1__6|SGB10068` (length 900)\n",
      "Found marker `UniRef90_P30130|1__4|SGB10068` (length 550)\n",
      "Found marker `UniRef90_UPI00063D0CB9|1__8|SGB10068` (length 1150)\n",
      "Found marker `UniRef90_A0A377LLX4|1__4|SGB10068` (length 550)\n",
      "Found marker `UniRef90_P40876|1__4|SGB10068` (length 600)\n",
      "Found marker `UniRef90_P77357|1__8|SGB10068` (length 1200)\n",
      "Found marker `UniRef90_P77268|1__6|SGB10068` (length 900)\n",
      "Found marker `UniRef90_P77416|1__9|SGB10068` (length 1350)\n",
      "Found marker `UniRef90_P76017|16__19|SGB10068` (length 550)\n",
      "Found marker `UniRef90_P76210|1__5|SGB10068` (length 750)\n",
      "Found marker `UniRef90_P77316|1__8|SGB10068` (length 1200)\n",
      "Found marker `UniRef90_P77626|1__3|SGB10068` (length 450)\n",
      "Found marker `UniRef90_P75937|3__6|SGB10068` (length 500)\n",
      "Found marker `UniRef90_P0AA51|1__6|SGB10068` (length 900)\n",
      "Found marker `UniRef90_P76204|1__4|SGB10068` (length 600)\n",
      "Found marker `UniRef90_B7UNN0|3__6|SGB10068` (length 550)\n",
      "Found marker `UniRef90_Q46808|7__10|SGB10068` (length 550)\n",
      "Found marker `UniRef90_Q46814|1__4|SGB10068` (length 550)\n",
      "Found marker `UniRef90_P77294|1__3|SGB10068` (length 450)\n",
      "Found marker `UniRef90_P0C8J9|4__8|SGB10068` (length 700)\n",
      "Found marker `UniRef90_P39381|1__7|SGB10068` (length 1050)\n",
      "Found marker `UniRef90_P76230|4__9|SGB10068` (length 850)\n",
      "Found marker `UniRef90_A0A066RI01|2__6|SGB10068` (length 650)\n",
      "Found marker `UniRef90_P0CI32|1__5|SGB10068` (length 750)\n",
      "Found marker `UniRef90_P77329|1__6|SGB10068` (length 850)\n",
      "Found marker `UniRef90_P76052|1__9|SGB10068` (length 1350)\n",
      "Found marker `UniRef90_Q8XAY7|2__6|SGB10068` (length 650)\n",
      "Found marker `UniRef90_P45757|1__5|SGB10068` (length 750)\n",
      "Found marker `UniRef90_P0ABD2|1__10|SGB10068` (length 1500)\n",
      "Found marker `UniRef90_P75715|1__6|SGB10068` (length 900)\n",
      "Found marker `UniRef90_P13656|13__17|SGB10068` (length 700)\n",
      "Found marker `UniRef90_P77468|1__17|SGB10068` (length 2550)\n",
      "Found marker `UniRef90_P25960|1__5|SGB10068` (length 750)\n",
      "Found marker `UniRef90_P75859|1__3|SGB10068` (length 450)\n",
      "Found marker `UniRef90_P75748|1__4|SGB10068` (length 550)\n",
      "Found marker `UniRef90_P76150|1__7|SGB10068` (length 1050)\n",
      "Found marker `UniRef90_P37641|3__6|SGB10068` (length 550)\n",
      "Found marker `UniRef90_P39165|1__7|SGB10068` (length 1000)\n",
      "Found marker `UniRef90_P58203|1__5|SGB10068` (length 700)\n",
      "Found marker `UniRef90_P09348|2__5|SGB10068` (length 550)\n",
      "Found marker `UniRef90_P39267|2__5|SGB10068` (length 550)\n",
      "Found marker `UniRef90_C8U5G9|1__3|SGB10068` (length 450)\n",
      "Found marker `UniRef90_P46139|2__8|SGB10068` (length 1000)\n",
      "Found marker `UniRef90_P0DM85|6__10|SGB10068` (length 650)\n",
      "Found marker `UniRef90_P75892|1__5|SGB10068` (length 700)\n",
      "Found marker `UniRef90_P77747|4__7|SGB10068` (length 550)\n",
      "Found marker `UniRef90_P76129|2__7|SGB10068` (length 800)\n",
      "Found marker `UniRef90_Q8XED7|10__13|SGB10068` (length 550)\n",
      "Found marker `UniRef90_P77493|2__6|SGB10068` (length 700)\n",
      "Found marker `UniRef90_P69832|6__9|SGB10068` (length 550)\n",
      "Found marker `UniRef90_P38055|3__8|SGB10068` (length 800)\n",
      "Found marker `UniRef90_P21363|1__3|SGB10068` (length 450)\n",
      "Found marker `UniRef90_P05704|1__10|SGB10068` (length 1500)\n",
      "Found marker `UniRef90_P75906|1__5|SGB10068` (length 700)\n",
      "Found marker `UniRef90_P76251|3__7|SGB10068` (length 700)\n",
      "Found marker `UniRef90_Q320T0|6__9|SGB10068` (length 550)\n",
      "Found marker `UniRef90_P46133|2__10|SGB10068` (length 1300)\n",
      "Found marker `UniRef90_P04846|2__5|SGB10068` (length 550)\n",
      "Found marker `UniRef90_P0AEL9|1__4|SGB10068` (length 600)\n",
      "Found marker `UniRef90_P45759|1__4|SGB10068` (length 550)\n",
      "Found marker `UniRef90_P41441|1__7|SGB10068` (length 1050)\n",
      "Found marker `UniRef90_P77249|1__4|SGB10068` (length 600)\n",
      "Found marker `UniRef90_P77454|3__6|SGB10068` (length 550)\n",
      "Found marker `UniRef90_P39414|4__7|SGB10068` (length 500)\n",
      "Found marker `UniRef90_P71229|1__13|SGB10068` (length 1950)\n",
      "Found marker `UniRef90_P31436|1__4|SGB10068` (length 550)\n",
      "Found marker `UniRef90_P77437|1__10|SGB10068` (length 1500)\n",
      "Found marker `UniRef90_A0A192EQ31|1__4|SGB10068` (length 600)\n",
      "Found marker `UniRef90_J7QR22|1__3|SGB10068` (length 450)\n",
      "Found marker `UniRef90_P77423|1__3|SGB10068` (length 450)\n",
      "Found marker `UniRef90_P75857|9__17|SGB10068` (length 1300)\n",
      "Found marker `UniRef90_Q83R14|3__6|SGB10068` (length 500)\n",
      "Found marker `UniRef90_J7QSW1|1__5|SGB10068` (length 750)\n",
      "Found marker `UniRef90_P77280|1__6|SGB10068` (length 900)\n",
      "Found marker `UniRef90_C6EHJ6|1__4|SGB10068` (length 600)\n",
      "Found marker `UniRef90_P77308|1__6|SGB10068` (length 900)\n",
      "Found marker `UniRef90_P45762|1__6|SGB10068` (length 900)\n",
      "Found marker `UniRef90_P45758|1__10|SGB10068` (length 1450)\n",
      "Found marker `UniRef90_P32140|2__8|SGB10068` (length 1000)\n",
      "Found marker `UniRef90_P77393|1__5|SGB10068` (length 750)\n",
      "Found marker `UniRef90_P07018|1__10|SGB10068` (length 1500)\n",
      "Found marker `UniRef90_P33924|1__17|SGB10068` (length 2550)\n",
      "Found marker `UniRef90_P75858|1__7|SGB10068` (length 1050)\n",
      "Found marker `UniRef90_P38052|1__3|SGB10068` (length 450)\n",
      "Found marker `UniRef90_P76017|7__14|SGB10068` (length 1100)\n",
      "Found marker `UniRef90_P52600|1__5|SGB10068` (length 700)\n",
      "Found marker `UniRef90_P75691|4__7|SGB10068` (length 550)\n",
      "Found marker `UniRef90_P77333|1__6|SGB10068` (length 900)\n",
      "Found marker `UniRef90_P33340|1__6|SGB10068` (length 900)\n",
      "Found marker `UniRef90_P77704|1__5|SGB10068` (length 750)\n",
      "Found marker `UniRef90_P17583|1__7|SGB10068` (length 1050)\n",
      "Found marker `UniRef90_P32137|3__7|SGB10068` (length 650)\n",
      "Found marker `UniRef90_P76573|1__3|SGB10068` (length 450)\n",
      "Found marker `UniRef90_P76518|1__7|SGB10068` (length 1050)\n",
      "Found marker `UniRef90_P45756|3__9|SGB10068` (length 1000)\n",
      "Found marker `UniRef90_P39374|1__5|SGB10068` (length 750)\n",
      "Found marker `UniRef90_P32701|2__10|SGB10068` (length 1300)\n",
      "Found marker `UniRef90_P77588|1__6|SGB10068` (length 900)\n",
      "Found marker `UniRef90_P69435|1__11|SGB10068` (length 1600)\n",
      "Found marker `UniRef90_P76773|1__4|SGB10068` (length 550)\n",
      "Found marker `UniRef90_P32136|1__8|SGB10068` (length 1150)\n",
      "Found marker `UniRef90_P77561|4__15|SGB10068` (length 1750)\n",
      "Found marker `UniRef90_P30855|17__23|SGB10068` (length 1000)\n",
      "Found marker `UniRef90_P32139|1__6|SGB10068` (length 900)\n",
      "Found marker `UniRef90_P30855|6__14|SGB10068` (length 1250)\n",
      "Found marker `UniRef90_P75908|6__9|SGB10068` (length 550)\n",
      "Found marker `UniRef90_P37627|6__9|SGB10068` (length 500)\n",
      "Found marker `UniRef90_P75908|1__4|SGB10068` (length 550)\n"
     ]
    }
   ],
   "source": [
    "metaphlan_genes = extract_from_metaphlan(metaphlan_pkl_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "408be46b-1915-4c03-9f39-ff57e0b2e741",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write each individual MetaPhlAn gene to fasta.\n",
    "metaphlan_index = []\n",
    "for metaphlan_gene in metaphlan_genes:\n",
    "    record = SeqRecord(\n",
    "        seq=metaphlan_gene.seq,\n",
    "        id=metaphlan_gene.id,\n",
    "        description=f\"MLST {taxon_label} marker gene version {metaphlan_version_id}\"\n",
    "    )\n",
    "\n",
    "    uniref_id, middle_tag, sgb_id = metaphlan_gene.id.split(\"|\")\n",
    "    # Turn this into something posix-friendly.\n",
    "    filename = f'{uniref_id}-{middle_tag}-{sgb_id}.fasta'\n",
    "    with open(gene_basedir / filename, 'wt') as f:\n",
    "        SeqIO.write([record], f, 'fasta')\n",
    "    \n",
    "    metaphlan_index.append({'Name': f'{uniref_id}-{middle_tag}-{sgb_id}', 'Fasta': filename, 'Metadata': f'MetaPhlAn:{metaphlan_version_id}'})\n",
    "\n",
    "metaphlan_index = pd.DataFrame(metaphlan_index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "33ca7ce7-c4b2-434c-ab14-74b43841e283",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Fasta</th>\n",
       "      <th>Metadata</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>UniRef90_P75933-1__4-SGB10068</td>\n",
       "      <td>UniRef90_P75933-1__4-SGB10068.fasta</td>\n",
       "      <td>MetaPhlAn:mpa_vJun23_CHOCOPhlAnSGB_202307</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>UniRef90_Q0T2M6-4__9-SGB10068</td>\n",
       "      <td>UniRef90_Q0T2M6-4__9-SGB10068.fasta</td>\n",
       "      <td>MetaPhlAn:mpa_vJun23_CHOCOPhlAnSGB_202307</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>UniRef90_P33916-5__8-SGB10068</td>\n",
       "      <td>UniRef90_P33916-5__8-SGB10068.fasta</td>\n",
       "      <td>MetaPhlAn:mpa_vJun23_CHOCOPhlAnSGB_202307</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>UniRef90_P25718-1__4-SGB10068</td>\n",
       "      <td>UniRef90_P25718-1__4-SGB10068.fasta</td>\n",
       "      <td>MetaPhlAn:mpa_vJun23_CHOCOPhlAnSGB_202307</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>UniRef90_Q8XAU9-1__3-SGB10068</td>\n",
       "      <td>UniRef90_Q8XAU9-1__3-SGB10068.fasta</td>\n",
       "      <td>MetaPhlAn:mpa_vJun23_CHOCOPhlAnSGB_202307</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>UniRef90_P32139-1__6-SGB10068</td>\n",
       "      <td>UniRef90_P32139-1__6-SGB10068.fasta</td>\n",
       "      <td>MetaPhlAn:mpa_vJun23_CHOCOPhlAnSGB_202307</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>UniRef90_P30855-6__14-SGB10068</td>\n",
       "      <td>UniRef90_P30855-6__14-SGB10068.fasta</td>\n",
       "      <td>MetaPhlAn:mpa_vJun23_CHOCOPhlAnSGB_202307</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>UniRef90_P75908-6__9-SGB10068</td>\n",
       "      <td>UniRef90_P75908-6__9-SGB10068.fasta</td>\n",
       "      <td>MetaPhlAn:mpa_vJun23_CHOCOPhlAnSGB_202307</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>UniRef90_P37627-6__9-SGB10068</td>\n",
       "      <td>UniRef90_P37627-6__9-SGB10068.fasta</td>\n",
       "      <td>MetaPhlAn:mpa_vJun23_CHOCOPhlAnSGB_202307</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>UniRef90_P75908-1__4-SGB10068</td>\n",
       "      <td>UniRef90_P75908-1__4-SGB10068.fasta</td>\n",
       "      <td>MetaPhlAn:mpa_vJun23_CHOCOPhlAnSGB_202307</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>136 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               Name                                 Fasta  \\\n",
       "0     UniRef90_P75933-1__4-SGB10068   UniRef90_P75933-1__4-SGB10068.fasta   \n",
       "1     UniRef90_Q0T2M6-4__9-SGB10068   UniRef90_Q0T2M6-4__9-SGB10068.fasta   \n",
       "2     UniRef90_P33916-5__8-SGB10068   UniRef90_P33916-5__8-SGB10068.fasta   \n",
       "3     UniRef90_P25718-1__4-SGB10068   UniRef90_P25718-1__4-SGB10068.fasta   \n",
       "4     UniRef90_Q8XAU9-1__3-SGB10068   UniRef90_Q8XAU9-1__3-SGB10068.fasta   \n",
       "..                              ...                                   ...   \n",
       "131   UniRef90_P32139-1__6-SGB10068   UniRef90_P32139-1__6-SGB10068.fasta   \n",
       "132  UniRef90_P30855-6__14-SGB10068  UniRef90_P30855-6__14-SGB10068.fasta   \n",
       "133   UniRef90_P75908-6__9-SGB10068   UniRef90_P75908-6__9-SGB10068.fasta   \n",
       "134   UniRef90_P37627-6__9-SGB10068   UniRef90_P37627-6__9-SGB10068.fasta   \n",
       "135   UniRef90_P75908-1__4-SGB10068   UniRef90_P75908-1__4-SGB10068.fasta   \n",
       "\n",
       "                                      Metadata  \n",
       "0    MetaPhlAn:mpa_vJun23_CHOCOPhlAnSGB_202307  \n",
       "1    MetaPhlAn:mpa_vJun23_CHOCOPhlAnSGB_202307  \n",
       "2    MetaPhlAn:mpa_vJun23_CHOCOPhlAnSGB_202307  \n",
       "3    MetaPhlAn:mpa_vJun23_CHOCOPhlAnSGB_202307  \n",
       "4    MetaPhlAn:mpa_vJun23_CHOCOPhlAnSGB_202307  \n",
       "..                                         ...  \n",
       "131  MetaPhlAn:mpa_vJun23_CHOCOPhlAnSGB_202307  \n",
       "132  MetaPhlAn:mpa_vJun23_CHOCOPhlAnSGB_202307  \n",
       "133  MetaPhlAn:mpa_vJun23_CHOCOPhlAnSGB_202307  \n",
       "134  MetaPhlAn:mpa_vJun23_CHOCOPhlAnSGB_202307  \n",
       "135  MetaPhlAn:mpa_vJun23_CHOCOPhlAnSGB_202307  \n",
       "\n",
       "[136 rows x 3 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metaphlan_index"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f9d54d06-37bb-4813-902f-1aebb953691e",
   "metadata": {},
   "source": [
    "# Write the index files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "65d69d6e-be12-4215-867e-32d2a025e32b",
   "metadata": {},
   "outputs": [],
   "source": [
    "concat_index = pd.concat([mlst_index, metaphlan_index])\n",
    "concat_index.to_csv(gene_basedir / \"marker_seed_index.tsv\", sep='\\t', index=False, header=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "chronostrain",
   "language": "python",
   "name": "chronostrain"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
